image/golem_blender_app/golem_blender_app/commands/utils.py

```diff
@@ -21,53 +24,127 @@ def get_db_connection(work_dir: Path):
+            'VALUES (?, ?)',
+            ((x, None) for x in range(subtasks_count)),
         )
 
 
-def update_subtask(
+def start_subtask(
         db,
         subtask_num: int,
-        status: SubtaskStatus,
-        unique_id: Optional[str] = None) -> None:
+        subtask_id: str,
+) -> None:
+    status = get_subtasks_statuses(db, [subtask_num])
+    if status and subtask_num in status:
+        if not status[subtask_num][0].is_computable():
+            raise RuntimeError(f"Subtask {subtask_num} already started")
+
     with db:
         db.execute(
-            'UPDATE subtask_status SET status = ? WHERE num = ?',
-            (status.value, subtask_num),
+            'INSERT INTO history(id, num, status) '
+            'VALUES (?, ?, ?)',
+            (subtask_id, subtask_num, SubtaskStatus.COMPUTING.value)
+        )
+        db.execute(
+            'UPDATE subtasks '
+            'SET id = ? '
+            'WHERE num = ?',
+            (subtask_id, subtask_num)
+        )
+
+
+def discard_subtask(
+        db,
+        subtask_id: str,
+) -> None:
+    with db:
+        db.execute(
+            'UPDATE history '
+            'SET status = ? '
+            'WHERE id = ? ',
+            (SubtaskStatus.ABORTED.value,
+             subtask_id)
+        )
+        db.execute(
+            'UPDATE subtasks '
+            'SET id = NULL '
+            'WHERE id = ?',
+            (subtask_id,)
+        )
+
+
+def update_subtask_status(
+        db,
+        subtask_id: str,
+        status: SubtaskStatus
+) -> None:
+    with db:
+        db.execute(
+            'UPDATE history '
+            'SET status = ? '
+            'WHERE id = ?',
+            (status.value, subtask_id),
         )
-        if unique_id:
-            db.execute(
-                'UPDATE subtask_status SET unique_id = ? WHERE num = ?',
-                (unique_id, subtask_num),
-            )
 
 
 def get_subtasks_statuses(
         db,
-        nums: List[int]) -> Dict[int, Tuple[SubtaskStatus, str]]:
+        nums: List[int]
+) -> Dict[int, Tuple[SubtaskStatus, str]]:
     set_format = ','.join(['?'] * len(nums))
+
     cursor = db.cursor()
     cursor.execute(
-        "SELECT num, status, unique_id FROM subtask_status WHERE num IN "
-        f"({set_format})",
-        (*nums,),
-    )
-    values = cursor.fetchall()
+        'SELECT S.num, H.status, S.id '
+        'FROM subtasks S '
+        'LEFT JOIN history H ON (H.id = S.id) '
+        f'WHERE S.num IN ({set_format})',
+        (*nums,))
+
     return {
-        v[0]: (SubtaskStatus(v[1]), v[2]) for v in values
+        row[0]: (SubtaskStatus(row[1]), row[2]) for row in cursor.fetchall()
     }
 
 
 def get_next_pending_subtask(db) -> Optional[int]:
     cursor = db.cursor()
     cursor.execute(
-        'SELECT num FROM subtask_status WHERE status = ? LIMIT 1',
-        (SubtaskStatus.PENDING.value,)
+        'SELECT S.num '
+        'FROM subtasks S '
+        'WHERE NOT EXISTS ('
```

Comment:
Wouldn't `WHERE S.id IS NULL` suffice here?

Reply:
I'd have to replace `update_subtask_status` with specialized functions to make sure that `id` is set to `NULL` when required. Thinking of it, I don't like the `update_subtask_status` approach at all and will refactor it.